File: lib/php/examples/06_disable_inactive_peers.php
#!/usr/bin/env php
<?php
/**
 * 06_disable_inactive_peers — Disable peers with no handshake in the last N days (hygiene cron).
 *
 * Usage:
 *   WG_BASE_URL=http://... WG_ADMIN_USER=... WG_API_TOKEN=... php examples/06_disable_inactive_peers.php [days]
 */

declare(strict_types=1);

require_once __DIR__ . '/../vendor/autoload.php';

use WgPortal\Client\WgPortalClient;

function requireEnv(string $key): string
{
    $value = getenv($key);
    if ($value === false || $value === '') {
        fwrite(STDERR, "Missing required env var: {$key}\n");
        exit(1);
    }
    return $value;
}

$inactiveDays = isset($argv[1]) ? (int) $argv[1] : 30;
$cutoff = (new DateTimeImmutable())->sub(new DateInterval("P{$inactiveDays}D"));
echo "Checking peers inactive since {$cutoff->format('Y-m-d')} ...\n\n";

$client = new WgPortalClient(
    requireEnv('WG_BASE_URL'),
    requireEnv('WG_ADMIN_USER'),
    requireEnv('WG_API_TOKEN')
);

$interfaces = $client->interfaces()->getAll();
$disabledCount = 0;

foreach ($interfaces as $iface) {
    $peers = $client->peers()->getByInterface($iface['Identifier']);
    foreach ($peers as $peer) {
        if ($peer['Disabled'] ?? false) {
            continue;
        }

        try {
            $metrics = $client->metrics()->byPeer($peer['Identifier']);
        } catch (\Throwable $e) {
            continue;
        }

        $shouldDisable = false;
        $reason = '';

        if (empty($metrics['LastHandshake'])) {
            echo "  Disabling {$peer['Identifier']} (never connected)\n";
            $shouldDisable = true;
            $reason = "No handshake (never connected, check > {$inactiveDays} days)";
        } else {
            try {
                $hs = new DateTimeImmutable($metrics['LastHandshake']);
                if ($hs < $cutoff) {
                    $last = $metrics['LastHandshake'];
                    echo "  Disabling {$peer['Identifier']} (last handshake: {$last})\n";
                    $shouldDisable = true;
                    $reason = "No handshake since {$last} (check > {$inactiveDays} days)";
                }
            } catch (\Throwable $e) {
                // can't parse date, skip
            }
        }

        if ($shouldDisable) {
            $client->peers()->update($peer['Identifier'], array_merge($peer, [
                'Disabled' => true,
                'DisabledReason' => $reason,
            ]));
            $disabledCount++;
        }
    }
}

echo "\nDisabled {$disabledCount} peer(s)\n";

File: lib/rust/examples/06_disable_inactive_peers.rs
/// 06_disable_inactive_peers — Disable peers with no handshake in the last N days (hygiene cron).
///
/// Usage:
///   WG_BASE_URL=http://... WG_ADMIN_USER=... WG_API_TOKEN=... cargo run --example 06_disable_inactive_peers [days]
use std::env;

use wg_portal_client::{WgPeer, WgPortalClient};

fn require_env(key: &str) -> String {
    env::var(key).unwrap_or_else(|_| {
        eprintln!("Missing required env var: {key}");
        std::process::exit(1);
    })
}

fn main() {
    let client = WgPortalClient::new(
        &require_env("WG_BASE_URL"),
        &require_env("WG_ADMIN_USER"),
        &require_env("WG_API_TOKEN"),
    );

    let args: Vec<String> = env::args().collect();
    let inactive_days: i64 = args.get(1).and_then(|s| s.parse().ok()).unwrap_or(30);
    let cutoff = chrono::Utc::now() - chrono::Duration::days(inactive_days);
    println!("Checking peers inactive since {} ...\n", cutoff.format("%Y-%m-%d"));

    let interfaces = client.interfaces().get_all().unwrap();
    let mut disabled_count = 0;

    for iface in &interfaces {
        let peers = client.peers().get_by_interface(&iface.identifier).unwrap();
        for peer in &peers {
            if peer.disabled.unwrap_or(false) {
                continue;
            }

            let last_handshake = match client.metrics().by_peer(&peer.identifier) {
                Ok(m) => m.last_handshake,
                Err(_) => continue,
            };

            let should_disable = match last_handshake {
                None => {
                    println!(
                        "  Disabling {} (never connected)",
                        peer.identifier
                    );
                    true
                }
                Some(ref hs) => {
                    if let Ok(dt) = chrono::DateTime::parse_from_rfc3339(hs) {
                        if dt < cutoff {
                            println!(
                                "  Disabling {} (last handshake: {hs})",
                                peer.identifier
                            );
                            true
                        } else {
                            false
                        }
                    } else {
                        false
                    }
                }
            };

            if should_disable {
                client
                    .peers()
                    .update(
                        &peer.identifier,
                        &WgPeer {
                            disabled: Some(true),
                            disabled_reason: Some(format!(
                                "No handshake since {last_handshake:?} (check > {inactive_days} days)"
                            )),
                            ..peer.clone()
                        },
                    )
                    .unwrap();
                disabled_count += 1;
            }
        }
    }

    println!("\nDisabled {disabled_count} peer(s)");
}
